File: wo/WO-M3-SPEC-KERNEL/guardian-live-20260825/path_guard.py
"""Path guard — symlink-safe containment checks (stdlib only).

The realpath primitive the rest of the governance core builds on, so a symlink
or ".." cannot smuggle a path past a prefix check. Imports nothing from the
lykoi package.
"""
from __future__ import annotations

import os


def is_within(path: str, base: str) -> bool:
    """True if ``path`` resolves to ``base`` or somewhere beneath it (realpath-safe)."""
    try:
        real = os.path.realpath(path)
        base_real = os.path.realpath(base)
    except OSError:
        return True  # unresolvable -> treat as inside (fail closed)
    return real == base_real or real.startswith(base_real + os.sep)


def classify(path: str, deny_zones) -> str:
    """Return "deny" if ``path`` resolves into any deny zone, else "allow"."""
    return "deny" if any(is_within(path, base) for base in deny_zones) else "allow"

File: packages/lykoi-kernel/src/path-guard.ts
import { realpathSync } from 'node:fs'
import { sep } from 'node:path'

export function isWithin(path: string, base: string): boolean {
  let real: string
  let baseReal: string
  try {
    real = realpathSync(path)
    baseReal = realpathSync(base)
  } catch {
    return true // unresolvable -> treat as inside (fail closed)
  }
  return real === baseReal || real.startsWith(baseReal + sep)
}

export function classify(path: string, denyZones: readonly string[]): 'deny' | 'allow' {
  return denyZones.some((base) => isWithin(path, base)) ? 'deny' : 'allow'
}
